pkg/plugin/plugin.go

```diff
@@ -118,15 +122,50 @@ func makeAutoscaleEnforcerPlugin(
 		}
 	}
 
+	// A note about event handling:
+	//
+	// Before returning from this function, we want to make sure that we're caught up to the watch
+	// events generated by initially reading the cluster state (i.e. the initial List()).
+	//
+	// Doing this is non-trivial, so we accomplish it in pieces:
+	//
+	// 1. Using watch.WatchModeSync to force queueing events *before* returning from creating the
+	//    watcher (note: and therefore, before any start to be handled); and
+	// 2. For each event created from the initial List(), increment a counter to track the number of
+	//    these events, and decrement it as events are handled.
+	//
+	// The initial state building is complete when the counter reaches zero, at which point we close
+	// the channel that this function will wait on.
+
+	var initEventsCount atomic.Int32
```

Comment:
Can you use [semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore#NewWeighted) here instead? 
 - Acquire(nil, n) when events are added
 - Release(1) when event is processed
 - Acquire(ctxWithTimeout, n) when waiting for events processing

Reply:
Seems simpler. Implemented it in caba29f33339c53e094935e66239ee375d01b26a but tbh I'm not sure if it's better. WDYT?

Reply:
If the `incEventCount` is left as a semaphore wrapper anyway, maybe instead revert to your old implementation, but wrap all in a struct synchronization primitive?

The reason I suggested semaphore is that I am wary of using raw channels and atomics in the code, where it feels like it should be a synchronization primitive.

Reply:
Made it a separate type in 521597e44affe70a50cccd1e0bf72d46fa500394 - LMK what you think.

Reply:
I imagined the implementation with initEventsCount hidden inside the new helper.

Also, you don't need incEventCount now

Reply:
> I imagined the implementation with initEventsCount hidden inside the new helper.

Hm, it's hard to make that kind of abstraction safe to use (because the counter isn't monotonic) -- maybe it'd be fine here, but I'd rather keep the logic inlined than make an unsafe abstraction for just this purpose.

So in its current state, incEventCount or equivalent is still needed (we can just directly add to the atomic; don't know whether that's better)